### ISSUE TITLE
virtual parentheses

### ISSUE THREAD

**Issue by fommil** [Fuco1/smartparens#470]
I'm one of the developers of https://github.com/ensime/ensime-server which, when used from emacs, is an external service that provides information about the scala code in the buffer.

Scala has a lot of "sugar" in the language which allows parenthesis to be omitted and the compiler is still ok with it (under the hood, the parenthesis - called "blocks" in the language spec - is inferred and added during one of the compile phases). However, this causes a nightmare for smartparens matching.

But in the ensime server, we actually know where all the blocks are, which means we know exactly which character should have a virtual parenthesis.

If we could provide a data structure of where the virtual parenthesis lives, do you think smartparens would be able to make use of the information to provide true sexp navigation?

**Comment by Fuco1** [comment]
Yes, I talked about this with someone from the python world.  It is "relatively" simple to plug different parsers into SP.  The API you need to implement is fairly thin, just two functions doing "give me next sexp after (before) point".  Everything else in SP is implemented in terms of that.

I might find some time over the weekend to write some specification of how exactly this works.

**Comment by fommil** [comment]
sweet!

We are thinking about trying to provide data structures that are "standard" (in the emacs sense, whatever that means). e.g. if there is some standard face overlay or something, that would be perfect. Ideally we'd want to use something that can change dynamically as the user is typing in the buffer. The ensime backend might only provide the information when the user saves the buffer, but we'd want emacs to at least progress the markers as the user enters or deletes text.

**Comment by PythonNut** [comment]
+1 I'd like to tinker with this as well. 

I think there's some room for an improved "sexp" in python, at least. I might be able to help if the documentation for the internals is improved. :)

**Comment by Fuco1** [comment]
There is the problem of the built-in parsers being completely ad-hoc right now.  I thought it would work by simply adding another step somewhere inbetween but it's going to break strings and some other hacks.

I've started the parser rewrite project several times in the past and it always lead to the same pile of nonsense :/
